main/src/ca/uwaterloo/flix/language/phase/unification/SetFormula.scala

```diff
@@ -163,6 +164,83 @@ object SetFormula {
     toCaseType(minimizedSetFormula, sym, m, loc)
   }
 
+  /**
+    * Returns `(lowerBound,upperBound)` if `tpe` is a case set, otherwise
+    * returns `None`. The bounds are only in terms of constants, cases that must
+    * always be present (lowerBound) and which cases that can maximally be
+    * present (upperBound).
+    * {{{
+    * boundsAnalysisType(s ++ <Expr.Cst>) =
+    *     Some(({Cst}, {Cst, Var, Not, And, Or, Xor}))
+    *
+    * boundsAnalysisType(s1 -- s2 -- <Expr.Xor, Expr.And> ++ <Expr.Cst, Expr.Not>) =
+    *     Some(({Cst, Not}, {Cst, Var, Not, Or}))
+    *
+    * boundsAnalysisType(Int32) = None
+    * }}}
+    */
+  private def boundsAnalysisType(tpe: Type, enm: TypedAst.RestrictableEnum): Option[(Set[Symbol.RestrictableCaseSym], Set[Symbol.RestrictableCaseSym])] = {
+    tpe.kind match {
+      case Kind.CaseSet(_) =>
+        val (m, setFormulaUniv) = mkEnv(List(tpe), SortedSet.from(enm.cases.keys))
+        val setFormula = fromCaseType(tpe, m, setFormulaUniv)
+        val (lower, upper) = boundsAnalysis(setFormula)(setFormulaUniv)
+        Some(cstToType(lower, m), cstToType(upper, m))
+      case _ => None
+    }
+  }
+
+  /**
+    * Converts a [[SetFormula]] set into a [[Symbol.RestrictableCaseSym]] set by
+    * the mappings in `m`.
+    */
+  private def cstToType(cst: Set[Int], m: Bimap[VarOrCase, Int]): Set[Symbol.RestrictableCaseSym] = {
+    cst.map(c => m.getBackward(c) match {
+      case Some(VarOrCase.Var(_)) => throw InternalCompilerException("Unexpected case set var in constant", SourceLocation.Unknown)
+      case Some(VarOrCase.Case(sym)) => sym
+      case None => throw InternalCompilerException("Unexpected empty case set constant mapping", SourceLocation.Unknown)
+    })
+  }
+
+  /**
+    * Returns `(lowerBound,upperBound)`. The bounds are only in terms of
+    * constants, cases that must always be present (lowerBound) and which cases
+    * that can maximally be present (upperBound).
+    * {{{
+    * boundsAnalysisType(s or Cst{1}) =
+    *     ({1}, {1, 2, ..., n})
+    *
+    * boundsAnalysisType((s1 and  s2 and Cst{4, 6}) or Cst{1, 3}) =
+    *     ({1, 3}, {1, 2, 3, 5, 7, 8, ..., n})
+    * }}}
+    */
+  private def boundsAnalysis(f: SetFormula)(implicit univ: Set[Int]): (Set[Int], Set[Int]) = {
+    val bot = SetFormula.Cst(Set.empty): SetFormula
+    val top = SetFormula.Cst(univ): SetFormula
+
+    // create all possible assignments of the free variables (exponentially many)
```

Comment:
Don't we already have a function that does this?

Reply:
`minimize` does this internally but not reusable. Ill try to refactor for reuse